### PR TITLE
Update `has_len_all_ranks` to use `Strategy.root_device`

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Same motivation as #11734
Part of #11922 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @rohitgr7 @ninginthecloud @otaj

### CODE REVIEW

File: pytorch_lightning/utilities/data.py [hunk 1]
```diff
@@ -121,7 +121,7 @@ def has_len_all_ranks(
     infinite dataloader."""
     try:
         local_length = len(dataloader)
-        total_length = training_type.reduce(torch.tensor(local_length).to(model.device), reduce_op="sum")
+        total_length = training_type.reduce(torch.tensor(local_length).to(training_type.root_device), reduce_op="sum")
```
Comment (carmocca) [inline pytorch_lightning/utilities/data.py:124]: Unrelated to the purpose of this PR but we could create this in the device directly

```suggestion
        total_length = training_type.reduce(torch.tensor(local_length, device=training_type.root_device), reduce_op="sum")
```

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/utilities/data.py [hunk 2]
```diff
@@ -121,7 +121,9 @@ def has_len_all_ranks(
     infinite dataloader."""
     try:
         local_length = len(dataloader)
-        total_length = training_type.reduce(torch.tensor(local_length).to(model.device), reduce_op="sum")
+        total_length = training_type.reduce(
+            torch.tensor(local_length, device=training_type.root_device), reduce_op="sum"
```
Comment (tchaton) [inline pytorch_lightning/utilities/data.py:125]: Should be renamed strategy.

Comment (kaushikb11) [reply pytorch_lightning/utilities/data.py:125]: @ananthsub

**Review by rohitgr7 — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Comment by carmocca** [comment]
The CPU test failures are real:

```python
FAILED trainer/properties/test_estimated_stepping_batches.py::test_num_stepping_batches_gpu[trainer_kwargs0-10]
FAILED trainer/properties/test_estimated_stepping_batches.py::test_num_stepping_batches_gpu[trainer_kwargs1-5]
FAILED trainer/properties/test_estimated_stepping_batches.py::test_num_stepping_batches_gpu[trainer_kwargs2-4]
FAILED trainer/properties/test_estimated_stepping_batches.py::test_num_stepping_batches_gpu[trainer_kwargs3-3]
FAILED trainer/properties/test_estimated_stepping_batches.py::test_num_stepping_batches_gpu[trainer_kwargs4-64]
```

cc @rohitgr7
